canals/pipeline/pipeline.py

```diff
@@ -346,22 +347,33 @@ def run(self, data: Dict[str, Any], debug: bool = False) -> Dict[str, Any]:
 
             component, inputs = inputs_buffer.popitem(last=False)  # FIFO
 
-            # if debug:
-            #     draw(deepcopy(self.graph), engine="graphviz", path=f"debug/step_{current_step}.jpg", running=component, queued=inputs_buffer.keys())
+            # Make sure it didn't run too many times already
+            self._check_max_loops(component)
 
             # **** IS IT MY TURN YET? ****
-            # Check if the node should be run or not
-            ready_to_run = self._ready_to_run(name=component, inputs=inputs, inputs_buffer=inputs_buffer)
+            # Check if the component should be run or not
+            action = self._calculate_action(name=component, inputs=inputs, inputs_buffer=inputs_buffer)
 
             # This component is missing data: let's put it back in the queue and wait.
-            if ready_to_run == "wait":
+            if action == "wait":
+                if not inputs_buffer:
+                    # What if there are no components to wait for?
+                    raise PipelineRuntimeError(
+                        f"'{component}' is stuck waiting for input, but there are no other components to run. "
+                        "This is likely a Canals bug. Open an issue at https://github.com/deepset-ai/canals."
+                    )
+
                 inputs_buffer[component] = inputs
                 continue
 
             # This component did not receive the input it needs: it must be on a skipped branch. Let's not run it.
-            if ready_to_run == "skip":
+            if action == "skip":
                 self.graph.nodes[component]["visits"] += 1
-                inputs_buffer = self._skip_downstream_nodes(component=component, inputs_buffer=inputs_buffer)
+                inputs_buffer = self._skip_downstream_unvisited_nodes(component=component, inputs_buffer=inputs_buffer)
+                continue
+
+            if action == "remove":
+                # This component has no reason of being in the run queue, it's best if we remove it
```

Comment:
Just to clarify for the future:

```suggestion
                # This component has no reason of being in the run queue and we need to remove it. For example, this can happen to components that are connected to skipped branches of the pipeline. 
```